File: src/test/java/net/pterodactylus/sone/utils/OptionalsTest.java
package net.pterodactylus.sone.utils;

import java.util.Arrays;
import java.util.List;

import com.google.common.base.Optional;
import com.google.common.collect.FluentIterable;
import org.hamcrest.MatcherAssert;
import org.hamcrest.Matchers;
import org.junit.Test;

/**
 * Unit test for {@link Optionals}.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class OptionalsTest {

	private final Object object1 = new Object();
	private final Object object2 = new Object();
	private final Object object3 = new Object();

	@Test
	public void canCreateOptionals() {
		new Optionals();
	}

	@Test
	public void isPresentFiltersCorrectOptionals() {
		List<Optional<Object>> optionals = Arrays.asList(
				Optional.of(object1), Optional.absent(),
				Optional.of(object2), Optional.absent(),
				Optional.of(object3), Optional.absent()
		);
		List<Optional<Object>> filteredOptionals =
				FluentIterable.from(optionals).filter(Optionals.isPresent()).toList();
		MatcherAssert.assertThat(filteredOptionals, Matchers.contains(
				Optional.of(object1), Optional.of(object2), Optional.of(object3)));
	}

	@Test
	public void getReturnsCorrectValues() {
		List<Optional<Object>> optionals = Arrays.asList(
				Optional.of(object1),
				Optional.of(object2),
				Optional.of(object3)
		);
		List<Object> objects = FluentIterable.from(optionals).transform(Optionals.get()).toList();
		MatcherAssert.assertThat(objects, Matchers.contains(object1, object2, object3));
	}

}

File: src/test/kotlin/net/pterodactylus/sone/utils/OptionalsTest.kt
package net.pterodactylus.sone.utils

import com.google.common.base.Optional
import com.google.common.base.Optional.fromNullable
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.contains
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.nullValue
import org.junit.Test
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Test for [Optional] utils.
 */
class OptionalsTest {

	@Test
	fun `present optional can be transformed with let`() {
		val optional = Optional.of(1)
		assertThat(optional.let { it + 1 }, equalTo(2))
	}

	@Test
	fun `empty optional is transform to null with let`() {
		val optional = Optional.absent<Int>()
		assertThat(optional.let { it + 1 }, nullValue())
	}

	@Test
	fun `present optional can be processed with also`() {
		val called = AtomicBoolean(false)
		Optional.of(1).also { if (it == 1) called.set(true) }
		assertThat(called.get(), equalTo(true))
	}

	@Test
	fun `absent optional is not processed with also`() {
		val called = AtomicBoolean(false)
		Optional.absent<Int>().also { called.set(true) }
		assertThat(called.get(), equalTo(false))
	}

	@Test
	fun `1 as optional is correct optional`() {
		val optional = 1.asOptional()
		assertThat(optional.get(), equalTo(1))
	}

	@Test
	fun `null as optional is asent optional`() {
		val optional = null.asOptional()
		assertThat(optional.isPresent, equalTo(false))
	}

	@Test
	fun testMapPresent() {
		val originalList = listOf(1, 2, null, 3, null)
		assertThat(originalList.mapPresent { fromNullable(it) }, contains(1, 2, 3))
	}

}
